paddle/phi/api/lib/api_utils.h

```diff
@@ -49,12 +57,28 @@ inline std::shared_ptr<phi::SelectedRows> TensorToSelectedRows(
   return std::dynamic_pointer_cast<phi::SelectedRows>(tensor.impl());
 }
 
+inline paddle::optional<const phi::SelectedRows&> TensorToSelectedRows(
+    const paddle::optional<Tensor>& tensor) {
+  if (tensor) {
+    return {*std::dynamic_pointer_cast<phi::SelectedRows>(tensor->impl())};
+  }
+  return {paddle::none};
+}
+
 /* ----------------- for infer_meta --------------------- */
 
 inline phi::MetaTensor MakeMetaTensor(const phi::DenseTensor& tensor) {
   return phi::MetaTensor(tensor);
 }
 
+inline paddle::optional<phi::MetaTensor> MakeMetaTensor(
```

Comment:
Seems that sometimes the function returns "paddle::optional<T>", while othertimes it return "paddle::optional<T&>". May I ask what is the design here, like what determines the return type?